zug/util.hpp

```diff
@@ -89,29 +90,69 @@ constexpr decltype(auto) invoke_composition(Fns&& fns, Args&&... args)
 }
 
 template <typename Fn, typename... Fns>
-struct composed
+struct composed : std::tuple<Fn, Fns...>
 {
-    std::tuple<Fn, Fns...> fns;
+    using function_tuple = std::tuple<Fn, Fns...>;
+
+    template <typename ArgsTuple>
+    composed(ArgsTuple&& args)
+        : function_tuple{std::forward<ArgsTuple>(args)}
+    {}
```

Comment:
The problem with this constructor is that it shadows the copy constructor. You can avoid that by doing `enable_if` tricks, but in this particular case, I think we can just remove this one and bring in the constructors from tuple by doing:
```
using function_tuple::function_tuple;
```
(Also, I often call the alias for the base class `base_t`).

Reply:
Tsh, of course ... these are the sort of things that you don’t think about at 2am after a bunch of rakija!

Reply:
In fact we're using the move/copy constructor of `std::tuple` here (we've already created the tuple with the `tuple_cat`), so the using doesn't bring them into scope.

Reply:
> In fact we're using the move/copy constructor of std::tuple here (we've already created the tuple with the tuple_cat), so the using doesn't bring the desired ctor into scope.

AFAIK it does bring those into scope.

Reply:
We discussed this elsewhere, but here's an example that shows that the `tuple&&` and `const tuple&` ctors are still hidden:

https://godbolt.org/z/s0oTtF